### DISCUSSION TITLE
2.3.91 Upgrade Failing to 2.3.100 -  Rendering SLS 'base:ssl' failed: Jinja variable list object has no element 0

### DISCUSSION THREAD

**Discussion by stondino00** [Security-Onion-Solutions/securityonion#7054] [Unsupported Versions] [answered]
I'm seeing the same issue as others are posting about. Upgraded my manager to 2.3.100 and that went fine. None of my nodes have upgraded and are still on 91.

When I try running highstate manually, I get this error.

[ERROR ] Rendering exception occurred
Traceback (most recent call last):
File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 502, in render_jinja_tmpl
output = template.render(**decoded_context)
File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 1090, in render
self.environment.handle_exception()
File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 832, in handle_exception
reraise(*rewrite_traceback_stack(source=source))
File "/usr/lib/python3.6/site-packages/jinja2/_compat.py", line 28, in reraise
raise value.with_traceback(tb)
File "", line 41, in top-level template code
jinja2.exceptions.UndefinedError: list object has no element 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 261, in render_tmpl
output = render_str(tmplstr, context, tmplpath)
File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 509, in render_jinja_tmpl
raise SaltRenderError("Jinja variable {}{}".format(exc, out), buf=tmplstr)
salt.exceptions.SaltRenderError: Jinja variable list object has no element 0
[CRITICAL] Rendering SLS 'base:ssl' failed: Jinja variable list object has no element 0
local:
Data failed to compile:

Rendering SLS 'base:ssl' failed: Jinja variable list object has no element 0

[soup.log](https://github.com/Security-Onion-Solutions/securityonion/files/7979703/soup.log)

**Comment by TOoSmOotH** [comment]
This error is different and has to do with the SSL state. Please attach your soup log.

**Reply by TOoSmOotH** [reply]
Also.. Are you getting this error on the manager or one of the minions?

**Reply by stondino00** [reply]
This error is on the minions when I run the highstate command.  The manager appeared to upgrade successfully.  I ran soup a number of times like it wanted to upgrade that component and then it upgraded to .100 successfully.

**Reply by TOoSmOotH** [reply]
can you run sudo salt-call state.apply ssl on your manager?

**Reply by stondino00** [reply]
That command succeeded on the manager with 0 failures.

I have attached the soup log from one of the minions.

**Reply by TOoSmOotH** [reply]
You don;t run soup on the minions, only the manager. 

Can you tell me what you get when you do:  
CentOS: rpm -qa |grep salt
Ubuntu: dpkg --list |grep salt

**Reply by TOoSmOotH** [reply]
the above commands on a minion

**Reply by stondino00** [reply]
Here is what I get.

salt-minion-3004-1.el7.noarch
salt-3004-1.el7.noarch

**Reply by stondino00** [reply]
I just attached the soup log file from the manager.

**Reply by TOoSmOotH** [reply]
Was that from the manager or a minion?

**Reply by stondino00** [reply]
It was from a minion, the salt rpm command output

**Reply by TOoSmOotH** [reply]
ok thanks. Wanted to make sure salt upgraded ok.

**Reply by TOoSmOotH** [reply]
Can you try doing a sudo service salt-master restart on the manager then running so-checkin on one of the minions?

**Reply by stondino00** [reply]
[INFO    ] Loading fresh modules for state activity
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/clouds'
[INFO    ] Syncing clouds for environment 'base'
[INFO    ] Loading cache from salt://_clouds, for base
[INFO    ] Caching directory '_clouds/' for environment 'base'
[INFO    ] Syncing beacons for environment 'base'
[INFO    ] Loading cache from salt://_beacons, for base
[INFO    ] Caching directory '_beacons/' for environment 'base'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_beacons/zeek.py' to '/var/cache/salt/minion/extmods/beacons/zeek.py'
[INFO    ] Syncing modules for environment 'base'
[INFO    ] Loading cache from salt://_modules, for base
[INFO    ] Caching directory '_modules/' for environment 'base'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_modules/healthcheck.py' to '/var/cache/salt/minion/extmods/modules/healthcheck.py'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_modules/needs_restarting.py' to '/var/cache/salt/minion/extmods/modules/needs_restarting.py'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_modules/so.py' to '/var/cache/salt/minion/extmods/modules/so.py'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_modules/telegraf.py' to '/var/cache/salt/minion/extmods/modules/telegraf.py'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_modules/zeekctl.py' to '/var/cache/salt/minion/extmods/modules/zeekctl.py'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/states'
[INFO    ] Syncing states for environment 'base'
[INFO    ] Loading cache from salt://_states, for base
[INFO    ] Caching directory '_states/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/sdb'
[INFO    ] Syncing sdb for environment 'base'
[INFO    ] Loading cache from salt://_sdb, for base
[INFO    ] Caching directory '_sdb/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/grains'
[INFO    ] Syncing grains for environment 'base'
[INFO    ] Loading cache from salt://_grains, for base
[INFO    ] Caching directory '_grains/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/renderers'
[INFO    ] Syncing renderers for environment 'base'
[INFO    ] Loading cache from salt://_renderers, for base
[INFO    ] Caching directory '_renderers/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/returners'
[INFO    ] Syncing returners for environment 'base'
[INFO    ] Loading cache from salt://_returners, for base
[INFO    ] Caching directory '_returners/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/output'
[INFO    ] Syncing output for environment 'base'
[INFO    ] Loading cache from salt://_output, for base
[INFO    ] Caching directory '_output/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/utils'
[INFO    ] Syncing utils for environment 'base'
[INFO    ] Loading cache from salt://_utils, for base
[INFO    ] Caching directory '_utils/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/log_handlers'
[INFO    ] Syncing log_handlers for environment 'base'
[INFO    ] Loading cache from salt://_log_handlers, for base
[INFO    ] Caching directory '_log_handlers/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/executors'
[INFO    ] Syncing executors for environment 'base'
[INFO    ] Loading cache from salt://_executors, for base
[INFO    ] Caching directory '_executors/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/proxy'
[INFO    ] Syncing proxy for environment 'base'
[INFO    ] Loading cache from salt://_proxy, for base
[INFO    ] Caching directory '_proxy/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/engines'
[INFO    ] Syncing engines for environment 'base'
[INFO    ] Loading cache from salt://_engines, for base
[INFO    ] Caching directory '_engines/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/thorium'
[INFO    ] Syncing thorium for environment 'base'
[INFO    ] Loading cache from salt://_thorium, for base
[INFO    ] Caching directory '_thorium/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/serializers'
[INFO    ] Syncing serializers for environment 'base'
[INFO    ] Loading cache from salt://_serializers, for base
[INFO    ] Caching directory '_serializers/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/matchers'
[INFO    ] Syncing matchers for environment 'base'
[INFO    ] Loading cache from salt://_matchers, for base
[INFO    ] Caching directory '_matchers/' for environment 'base'
[INFO    ] Loading fresh modules for state activity
[INFO    ] Executing command 'yum' in directory '/root'
[INFO    ] Executing command rpm in directory '/root'
[INFO    ] Executing command 'yum' in directory '/root'
[ERROR   ] Rendering exception occurred
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 502, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/lib/python3.6/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 41, in top-level template code
jinja2.exceptions.UndefinedError: list object has no element 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 261, in render_tmpl
    output = render_str(tmplstr, context, tmplpath)
  File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 509, in render_jinja_tmpl
    raise SaltRenderError("Jinja variable {}{}".format(exc, out), buf=tmplstr)
salt.exceptions.SaltRenderError: Jinja variable list object has no element 0
[CRITICAL] Rendering SLS 'base:ssl' failed: Jinja variable list object has no element 0
local:
    Data failed to compile:



    Rendering SLS 'base:ssl' failed: Jinja variable list object has no element 0

**Reply by stondino00** [reply]
What should I try next?

**Reply by m0duspwnens** [reply]
On one of the minions where you are seeing `Rendering SLS 'base:ssl' failed: Jinja variable list object has no element 0`, can you run the following and post the output with sensitive information redacted?

This should return the value of your management interface.
`salt-call pillar.get host:mainint`

If the previous command returned properly, run the following. This should list all the interfaces on the node as well as IPs associated with them?
`salt-call grains.get ip_interfaces`

This should return the value of `/etc/pki/ca.crt`.
`salt-call mine.get so-manager_managersearch x509.get_pem_entries`

**Reply by stondino00** [reply]
Here is the output of all those commands.


[sohybridad@SO-Main ~]$ sudo salt-call pillar.get host:mainint
local:
    enp8s0f3

[sohybridad@SO-Main ~]$ sudo salt-call grains.get ip_interfaces
local:
    ----------
    bond0:
    docker0:
        - 172.17.0.1
    enp1s0f0:
    enp1s0f1:
    enp8s0f0:
    enp8s0f1:
    enp8s0f2:
    enp8s0f3:
        - 192.168.20.59
    ens5f0:
    ens5f1:
    lo:
        - 127.0.0.1
    veth1174833:
    veth14cdb3f:
    veth1a0e8a4:
    veth28a96c8:
    veth51501e5:
    veth5d70a6b:
    veth84d6daf:
    veth864113d:
    vethc14c1ab:
    vethcd88b5c:
    vethe4f165c:
    vethe691916:
    vethec3f93e:

[sohybridad@SO-Main ~]$ sudo salt-call mine.get so-manager_managersearch x509.get_pem_entries
local:
    ----------
    so-manager_managersearch:
        ----------
        /etc/pki/ca.crt:
            -----BEGIN CERTIFICATE-----
            MIIF1TCCA72gAwIBAgIIQ/Rf2bhWliowDQYJKoZIhvcNAQELBQAwSjELMAkGA1UE
            BhMCVVMxEzARBgNVBAMMClNPLU1hbmFnZXIxFzAVBgNVBAcMDlNhbHQgTGFrZSBD
            aXR5MQ0wCwYDVQQIDARVdGFoMB4XDTIwMTIwOTIxMDA0NloXDTMwMTIwNzIxMDA0
            NlowSjELMAkGA1UEBhMCVVMxEzARBgNVBAMMClNPLU1hbmFnZXIxFzAVBgNVBAcM
            DlNhbHQgTGFrZSBDaXR5MQ0wCwYDVQQIDARVdGFoMIICIjANBgkqhkiG9w0BAQEF
            AAOCAg8AMIICCgKCAgEA5XBeYhXyULd7jIoyhgJKD4OcPUlKyxjgqTw2NTFzayYB
            2T+y9r8C0Wt8i6QDe+DFdzybcvKM5RjZHJLVDlV33TFMXduON2nTkkLnvBs35IIt
            stwbYiq5VhirrbebtpXktjmVC50507mkNhMRncxyyfV+IgZLTVVGZnGyXqS9Sg13
            k3zvl91uB7g2q3xii2StcqNcZndLr8dwV8MwKdWEUVTvESAwfYlFHsQ6MhQc7uax
            wVsKnG2Ez/NYKblwup/nlrwhC2HKDF/O98aXaIs6+LfJ5K/ZYjZW6c+FFkQ5UA1s
            axqzMk096pcSjiYnMb2t9F+6/OYgBtOCY17Es24K5C4Kr9cJDWASn1iOxrpFb9ja
            ToHBkHqHQFyquZtZusdNNsx77eQN544ObEPiJw8EwBHu7iVHls8VT7L7uuaxqa8t
            s8OVnNcVVf33xxqgQG/4BTvdTRvVIF1cl3tLKMgtMOuUpHELy5LDGilCLo0C8vY3
            sAKpgx5+IlUyFdrCNNGV+BfFtwJfJwaptlKuJ5LPo8w5h2PWckNCtmiGpQy4cf36
            AKSzMPhlp0N5Rhv00VeQizAfU1/wUp/gYo7uDA0oF022BzaILwwAbcpnxCKj7oWT
            OyX2ihvubWL45g71g/5Zx93DoFgHxpJVBGGdQ5KfMk54ZMLSz9xl9sh3zlLs5KsC
            AwEAAaOBvjCBuzAPBgNVHRMBAf8EBTADAQH/MA4GA1UdDwEB/wQEAwIBBjAdBgNV
            HQ4EFgQUIgApf+3Q4OyP1E78LNg7vcR8jVYweQYDVR0jBHIwcIAUIgApf+3Q4OyP
            1E78LNg7vcR8jVahTqRMMEoxCzAJBgNVBAYTAlVTMRMwEQYDVQQDDApTTy1NYW5h
            Z2VyMRcwFQYDVQQHDA5TYWx0IExha2UgQ2l0eTENMAsGA1UECAwEVXRhaIIIQ/Rf
            2bhWliowDQYJKoZIhvcNAQELBQADggIBAAYSkjlJ+kcuUTBM6S/2lDR9IFhLTKLP
            3O28giAq99XLShPTwuitB7M1zkeZ9QStVomT6RA9j9HyNqHo0oAGhHcmPYpZgKtG
            VkRqWDhJ8mBk075sqYNFx7HX/5icPhQ1NdEhiObJMmqAtrn+mmxMPO6fG7CNzLb6
            4it2lMW+0ZF8R28RheqW1QPKv6wUCIxZycjUHTKBHFulOdDYfiX1x2QGO1WyT/Rt
            X1IoXa0GvremMN5EMb4astCORhVebJvkrU3muf6yPbsx9697HjK3aC4hM7nJOMeM
            +PhSiFNEM/ANrLZi+5ghz5GiqhWxP+0fb/N9OvcY44v9+3BfiQ31307VGQReVa5D
            bhgmRpCHfqvbHjIjby3w9Jllw8jmrsX1giIw6R55tNCO+hd9OXA6LXCxSWpimbRu
            BpxlaW9oaVtW381QuCwM+jhm0Dzc+zSsm/9E1ia93mRAW9GMR7X9CxoY2SVAxPnX
            B4FdtemMfOYzJ6T0EhBWiixMn4LcGqV4SfF2GFjyXlyqWcTZ7st7bEbs/iR5Yduj
            rKusAX8ihUtxQc6Sk956qfn5c/04WKsiBAKllTIDjnrNS5f50KDF2+h+dfpr/3IS
            PMQYDQjs5u9Z6lTz897POoPwQw2YOHCzmu1EsFXbLx2DudhPbzo8Oo/wttTJis5s
            kCb/hzJLWSlQ
            -----END CERTIFICATE-----

**Reply by m0duspwnens** [reply]
Everything looks fine there. I will try to replicate this is the lab, but in the meantime, can you restart the salt-minion service on `so-main` with `systemctl restart salt-minion`?

Once that is done, we will want to test the `ssl` state again with `salt-call state.apply ssl -linfo`. You may or may not get a message that resembles the following. If you do, just wait a bit and try again.
```
local:
    Data failed to compile:
----------
    The function "state.highstate" is running as PID
```

Thanks.

**Reply by stondino00** [reply]
Ok sounds good.  I've not had any issues upgrading previous versions until this one.

**Reply by stondino00** [reply]
Also I've tried rebooting all minions and manager and that didn't fix anything.

**Comment by m0duspwnens** [comment, marked as answer]
@stondino00 I was able to replicate and identify the issue. To resolve this, prior to waiting for us to release a hotfix, please make the following change. The issue has to do with a small code change to the `ssl` state and it not properly handling manager hostnames that contain uppercase characters.

In `/opt/so/saltstack/default/salt/ssl/init.sls`:
On line 26 you should see the following line: `{% set x509dict = salt['mine.get'](manager~'*', 'x509.get_pem_entries') %}`
Make the following modification: `{% set x509dict = salt['mine.get'](manager | lower~'*', 'x509.get_pem_entries') %}`

Once that is done, save and close the file. Then, on your nodes that were not able to run the `ssl` state, run `salt-call state.apply ssl -l info`. That should now succeed. Once it does, run `salt-call state.highstate -l info`.

**Reply by stondino00** [reply]
Awesome thanks so much!  That appears to have done it!

**Comment by influence2006** [comment]
Ran into this issue and did the suggest fix  @m0duspwnens recommended but i am getting zero activity form anything 
![image](https://user-images.githubusercontent.com/64173630/152201444-5f34d9a2-c4cd-4fa1-9704-22b77f7d05c5.png)
![image](https://user-images.githubusercontent.com/64173630/152201481-33128a3a-4d19-41d5-bffc-6a362d1a7ba8.png)

**Reply by influence2006** [reply]
@m0duspwnens  i dont seem to get any pki return 
salt-call mine.get so-manager_managersearch x509.get_pem_entries
local:
    ----------
this is all i get from the output

**Reply by m0duspwnens** [reply]
@influence2006 The command posted above was specific that user's manager hostname. From your manager, you would want to run `salt-call grains.get id`. From there, use that output and run the command `salt-call mine.get <manager_id> x509.get_pem_entries`.  That would return your ca.crt. This bug is only caused by a manager hostname that has uppercase characters.

**Reply by influence2006** [reply]
@m0duspwnens my manager has an uppercase name SecurityOnion

**Reply by influence2006** [reply]
@m0duspwnens I did happen tog et the pki/ca.crt with the right id thank you
but still same problem for hours
